clang/lib/Sema/SemaDeclCXX.cpp

```diff
@@ -9442,9 +9442,21 @@ bool SpecialMemberDeletionInfo::shouldDeleteForSubobjectCall(
 
   int DiagKind = -1;
 
-  if (SMOR.getKind() == Sema::SpecialMemberOverloadResult::NoMemberOrDeleted)
-    DiagKind = !Decl ? 0 : 1;
-  else if (SMOR.getKind() == Sema::SpecialMemberOverloadResult::Ambiguous)
+  if (SMOR.getKind() == Sema::SpecialMemberOverloadResult::NoMemberOrDeleted) {
+    if (CSM == Sema::CXXDefaultConstructor && Field &&
+        Field->getParent()->isUnion()) {
+      // [class.default.ctor]p2:
+      //   A defaulted default constructor for class X is defined as deleted if
+      //   - X is a union that has a variant member with a non-trivial default
+      //     constructor and no variant member of X has a default member
+      //     initializer
+      const auto *RD = cast<CXXRecordDecl>(Field->getParent());
+      if (!RD->hasInClassInitializer())
+        DiagKind = !Decl ? 0 : 1;
```

Comment:
So what is going on here?  this is the same value as its 'else' clause, and leaves DiagKind as -1 otherwise?  I might suggest making this a single 'if' and getting rid of the 'else' here.

Reply:
I have three different states I need to check, I can remove a check by setting `DiagKind` unconditionally and then check `if(RD->hasInClassInitializer())` instead and in that case set it back to `-1` but that does not express the intent as clearly. So I think what I have now is more maintainable although slightly more verbose.

Reply:
FWIW, I ran into exactly the same issue that Erich did in noticing that both branches do `DiagKind = !Decl ? 0 : 1;`.  That two of us ran into the same readability issue suggests that this should be modified to be more clear.

Reply:
You can refactor it like this: https://github.com/MitalAshok/llvm-project/commit/a119d89788c2c642b5a13ea00b2ea649b04a418e and I would suggest changing to an enum for readability: https://github.com/MitalAshok/llvm-project/commit/38ca82e1a55012e7d37fd8c6d049d1f00bb50835 (This can then be extracted out into a static function that returns the enum and each `DiagKind = ...` can turn into `return ...`, and the `if`/`else if` ladder can be dismantled)